contracts/0.4.24/Lido.sol

```diff
@@ -946,167 +905,78 @@ contract Lido is Versioned, StETHPermit, AragonApp {
-            }
-        }
-    }
-
-    function _transferTreasuryRewards(uint256 treasuryReward) internal {
-        address treasury = _treasury();
-        _transferShares(address(this), treasury, treasuryReward);
-        _emitTransferAfterMintingShares(treasury, treasuryReward);
-    }
-
-    /**
-     * @dev Gets the amount of Ether temporary buffered on this contract balance
-     */
+    /// @dev Get the amount of ether temporary buffered on this contract balance
     function _getBufferedEther() internal view returns (uint256) {
         return BUFFERED_ETHER_POSITION.getStorageUint256();
     }
 
+    /// @dev Set the amount of ether temporary buffered on this contract balance
     function _setBufferedEther(uint256 _newBufferedEther) internal {
         BUFFERED_ETHER_POSITION.setStorageUint256(_newBufferedEther);
     }
 
-    /// @dev Calculates and returns the total base balance (multiple of 32) of validators in transient state,
+    /// @dev Calculate and return the total base balance (multiple of 32) of validators in transient state,
     ///     i.e. submitted to the official Deposit contract but not yet visible in the CL state.
-    /// @return transient balance in wei (1e-18 Ether)
-    function _getTransientBalance() internal view returns (uint256) {
+    /// @return transient ether in wei (1e-18 Ether)
+    function _getTransientEther() internal view returns (uint256) {
         uint256 depositedValidators = DEPOSITED_VALIDATORS_POSITION.getStorageUint256();
         uint256 clValidators = CL_VALIDATORS_POSITION.getStorageUint256();
         // clValidators can never be less than deposited ones.
         assert(depositedValidators >= clValidators);
+
         return (depositedValidators - clValidators).mul(DEPOSIT_SIZE);
     }
 
-    /**
-     * @dev Gets the total amount of Ether controlled by the system
-     * @return total balance in wei
-     */
-    function _getTotalPooledEther() internal view returns (uint256) {
+    /// @dev Get the total amount of ether controlled by the protocol internally
+    /// (buffered + CL balance of StakingRouter controlled validators + transient)
+    function _getInternalEther() internal view returns (uint256) {
         return _getBufferedEther()
             .add(CL_BALANCE_POSITION.getStorageUint256())
-            .add(_getTransientBalance());
+            .add(_getTransientEther());
+    }
+
+    /// @dev Calculate the amount of ether controlled by external entities
+    function _getExternalEther(uint256 _internalEther) internal view returns (uint256) {
+        // TODO: cache external ether to storage
+        // to exchange 1 SLOAD in _getTotalPooledEther() 1 SSTORE in mintEE/burnEE
+        // _getTPE is super wide used
+        uint256 externalShares = EXTERNAL_SHARES_POSITION.getStorageUint256();
+        uint256 internalShares = _getTotalShares() - externalShares;
+        return externalShares.mul(_internalEther).div(internalShares);
```

Comment:
precision loss here can add up in `getPooledEthForShares()` and we get double loss on conversion. 

Probably, worth to redesign conversion functions in `StETH.sol` to use internal ETH instead.